src/Microsoft.ML.Core/Data/IEstimator.cs

```diff
@@ -16,13 +17,17 @@ namespace Microsoft.ML.Core.Data
     /// This is more relaxed than the proper <see cref="ISchema"/>, since it's only a subset of the columns,
     /// and also since it doesn't specify exact <see cref="ColumnType"/>'s for vectors and keys.
     /// </summary>
-    public sealed class SchemaShape
+    public sealed class SchemaShape : IReadOnlyList<SchemaShape.Column>
```

Comment:
>: IReadOnlyList<SchemaShape.Column> [](start = 36, length = 35)

This is not right! You can make this a `ReadOnlyList` with proper indexers and so forth. *Or* you could expose this `Columns` thing as you do below. You *CANNOT DO BOTH*.

Reply:
You're absolutely right. Thanks a lot for pointing this out.

---
In reply to: [239226291](https://github.com/dotnet/machinelearning/pull/1822#discussion_r239226291) [](ancestors = 239226291)